src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -1561,11 +1595,17 @@ static int client_parse_message(
         lease->next_server = message->siaddr;
         lease->address = message->yiaddr;
 
+        if (client->bootp)
+                lease->lifetime = USEC_INFINITY;
+
+        if (lease->server_address == 0 && !client->bootp)
+                return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(ENOMSG),
+                                             "received lease lacks server address, ignoring.");
```

Comment:
As far as I can see interfaces still go down with
```sh
host0: DHCP error: failed to get DHCP server IP address: No data available
host0: Failed
host0: State changed: configuring -> failed
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChange>
host0: The interface entered the failed state frequently, refusing to reconfigure it automatically.
host0: DHCPv4 client: STOPPED
host0: DHCPv4 client: State changed: bound -> stopped
```
when BOOTP replies come with no Server Identifier options. `bootpc` uses the `siaddr` field as far as I can tell because BOOTP is different from DHCP in that in BOOTP according to https://www.rfc-editor.org/rfc/rfc951#section-3
```
         siaddr  4       server IP address;
                         returned in bootreply by server.
```
while according to https://www.rfc-editor.org/rfc/rfc2131#section-2 in DHCP it's
```sh
   siaddr        4  IP address of next server to use in bootstrap;
                    returned in DHCPOFFER, DHCPACK by server.
```
(I haven't looked closely at that part yet though. I just took a look at what `bootpc` does)

Reply:
For the record `bootpd` sets the siaddr field in https://sources.debian.org/src/bootp/2.4.3-19.1/bootpd.c/#L1171-L1199 and `bootpc` process it in https://sources.debian.org/src/bootpc/0.64-9.1/bootpc.c/#L336-L337.

I have to admit I'm not sure why the BOOTP reply mentioned in https://github.com/systemd/systemd/issues/34885 came with no siaddr set. `bootpc` would have just extracted `0.0.0.0` and failed to process the Server Identifier option as far as I can tell.

Reply:
> the bootp server isn't really a bootp server, it does a 4-message exchange 

As far as I can see it does that when its DHCP mode is enabled. When it's off it doesn't process the DHCP message types.

> It should not be used as a baseline for BOOTP behavior

Agreed. The RFCs should be used for that. That being said given that the BOOTP reply mentioned in https://github.com/systemd/systemd/issues/34885 isn't RFC-compliant it's still useful to collect examples of BOOTP replies various BOOTP servers can send.

>  bootp server and bootpc client are both highly broken

I think at least the part of bootpc sending requests is OK in that it does exactly what the RFC says. Generally it seems to be hard to find a BOOTP implementation that isn't broken one way or another. If there is anything that isn't broken (and open-source) it would be great if you could share the pointers. It could be used to add an integration test for this feature for example.

Reply:
Before I forget the code handling BOOTP in dnsmasq can be found in https://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=blob_plain;f=src/rfc2131.c;hb=HEAD. I think it's quite useful.

(dnsmasq is already used in the systemd test suite so it's a plus)

Reply:
Here's how dnsmasq replies by default to BOOTP requests:
```sh
Dynamic Host Configuration Protocol
    Message type: Boot Reply (2)
    Hardware type: Ethernet (0x01)
    Hardware address length: 6
    Hops: 0
    Transaction ID: 0x24bf4ee4
    Seconds elapsed: 1
    Bootp flags: 0x0000 (Unicast)
        0... .... .... .... = Broadcast flag: Unicast
        .000 0000 0000 0000 = Reserved flags: 0x0000
    Client IP address: 0.0.0.0
    Your (client) IP address: 192.168.66.82
    Next server IP address: 192.168.66.81
    Relay agent IP address: 0.0.0.0
    Client MAC address: 11:22:33:44:55:66 (11:22:33:44:55:66)
    Client hardware address padding: 00000000000000000000
    Server host name not given
    Boot file name not given
    Magic cookie: DHCP
    Option: (1) Subnet Mask (255.255.255.240)
        Length: 4
        Subnet Mask: 255.255.255.240
    Option: (28) Broadcast Address (192.168.66.95)
        Length: 4
        Broadcast Address: 192.168.66.95
    Option: (3) Router
        Length: 4
        Router: 192.168.66.81
    Option: (255) End
        Option End: 255
    Padding: 0000000000000000000000000000000000000000000000000000000000000000000000000000000000
```
It's RFC-compliant and can be used to test the feature on the CI. The issue where network interfaces go down can be triggered with it.

Reply:
> > the bootp server isn't really a bootp server, it does a 4-message exchange
> 
> As far as I can see it does that when its DHCP mode is enabled. When it's off it doesn't process the DHCP message types.

I'm not sure what you mean by DHCP mode being enabled here. If you are just referring to the fact that does this when it sees a DHCP_DISCOVERY, then yes - let me clarify my statement. `bootpd` is more than just a bootp serer, it looks like it is trying to support as much of DHCP as it can, while still supporting bootp too.

> > bootp server and bootpc client are both highly broken
> 
> I think at least the part of bootpc sending requests is OK in that it does exactly what the RFC says. 

Agreed

> Generally it seems to be hard to find a BOOTP implementation that isn't broken one way or another. 

Agreed. The main things that I consider broken about `bootpd` are:

- It has no way to send replies to a global broadcast IP address
  - It probably should default to using 0.0.0.0 or 1.1.1.1 as the IP DST ADDR even if it is going to unicast the reply at the MAC layer, since it is reasonable to expect it to interoperate with `bootpc`, and that is sufficient to get `bootpc` to receive replies, despite its lack of using a raw socket
  - It incorrectly uses a subnet broadcast address when the BROADCAST REPLY option is sent. By definition a host can not know whether a destination IP address is a broadcast or not if it has a mix of 1's and 0's and the interface does not have an address & mask.

> If there is anything that isn't broken (and open-source) it would be great if you could share the pointers. It could be used to add an integration test for this feature for example.

I will update as soon as we get specifics from the customer. I've also been told that dhclient has bootp compatibility capability, but I haven't figured out a way to get it to behave that way. I haven't found a complete reference on its configuration options so yet.

Reply:
(Just to sync it with the issue to make it a bit easier (for me at least) to follow all the threads)

https://github.com/systemd/systemd/issues/34885#issuecomment-2506663049 is related to this thread.

Reply:
Semi-duplicating my comment from the issue b/c this is the appropriate place for it:

I've got a complete round-trip between the client and my `bootpd` server. I ran into this exact issue, lack of `server_address`. The client reported the same error, and did not configure the interface:
```
host0: DHCPv4 client: STARTED on ifindex 2
host0: DHCPv4 client is configured, acquiring DHCPv4 lease.
host0: DHCPv4 client: DISCOVER
host0: DHCPv4 client: State changed: initialization -> selecting
host0: DHCPv4 client: Received message from RAW socket, processing.
host0: DHCPv4 client: ACK
host0: DHCPv4 client: State changed: selecting -> bound
host0: DHCP error: failed to get DHCP server IP address: No data available
host0: Acquiring DHCPv4 lease.
```

Note that although it says `Acquiring DHCPv4 lease`, it did not configure the interface.

Here's a packet capture: [add_bootp-sendoption.cap.zip](https://github.com/user-attachments/files/17951582/add_bootp-sendoption.cap.zip)